internal/engine/wazevo/frontend/lower.go

```diff
@@ -3368,24 +3372,35 @@ func (c *Compiler) lowerCallIndirect(typeIndex, tableIndex uint32) {
 
 // memOpSetup inserts the bounds check and calculates the address of the memory operation (loads/stores).
 func (c *Compiler) memOpSetup(baseAddr ssa.Value, constOffset, operationSizeInBytes uint64) (address ssa.Value) {
+	address = ssa.ValueInvalid
 	builder := c.ssaBuilder
 
+	baseAddrID := baseAddr.ID()
 	ceil := constOffset + operationSizeInBytes
+	if known := c.getKnownSafeBound(baseAddrID); known.valid() {
+		// We reuse the calculated absolute address even if the bound is not known to be safe.
+		address = known.absoluteAddr
+		if ceil <= known.bound {
```

Comment:
As discussed on Slack, I struggle to be sure this is enough proof that the bound is safe. `ceil == know.bound` should be, but that's a missed opportunity.

Is `constOffset` never negative?

Reply:
constOffset is a "unsigned" 32-bit constant

Reply:
I mean, I know they're uint64, but how do "negative" offsets get encoded?

Are they not possible at this level? Is it not possible for address overflow to not be a trap (because it wraps around to a safe address)?

I'm probably displaying my ignorance of WASM here.
Sorry for my scepticism if so.

Reply:
no, if you take a look at the encoding of memory instructions in wasm spec, the offsets are encoded in **unsigned** 32-bit as I said ^^

Reply:
<img width="889" alt="image" src="https://github.com/tetratelabs/wazero/assets/13513977/fbd6cf1a-0cec-46a0-b366-d035137f11de">
https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/binary/instructions.html#memory-instructions

memarg's offset is not `s32`, but it is `u32, and is encoded by **unsigned** 32-bit leb128. FWIW, the logic you are commenting is exactly the same as the non-optimized bound check

Reply:
OK. I just always assumed that still allowed for the possibility of base+offset wrapping around in `uint32` math, and making an invalid access valid. But I haven't really read the spec, so I'll take your word for it.

Reply:
yeah ack this is tricky and thank you for your patience!

Reply:
Thank you for putting up with my ignorance and scepticism. 😁

I also found [this](https://github.com/WebAssembly/design/issues/584) which explains it as an explicit design choice to make sandboxing easier, which makes sense.